File: sdks/node/src/serializers/serializer.ts
/**
 * Pluggable codec for task arguments and results. Mirrors the cross-SDK
 * serializer contract so implementations can interoperate across languages.
 * The Rust core stores payloads as opaque bytes.
 */
export interface Serializer {
  serialize(value: unknown): Uint8Array;
  deserialize(bytes: Uint8Array): unknown;
  /**
   * Optional call-shaped encoding: wire serializers write the cross-SDK call
   * body `[args, kwargs]` (BINDING_CONTRACT.md) instead of a bare args array.
   * Results always use `serialize`/`deserialize`.
   */
  serializeCall?(args: unknown[]): Uint8Array;
  /** Inverse of {@link serializeCall}: returns handler-ready positional args. */
  deserializeCall?(bytes: Uint8Array): unknown[];
}

/** Encode task args via `serializeCall` when the serializer is call-aware. */
export function serializeCall(serializer: Serializer, args: unknown[]): Uint8Array {
  return serializer.serializeCall ? serializer.serializeCall(args) : serializer.serialize(args);
}

/** Decode a call payload into positional args, honoring call-aware serializers. */
export function deserializeCall(serializer: Serializer, bytes: Uint8Array): unknown[] {
  return serializer.deserializeCall
    ? serializer.deserializeCall(bytes)
    : (serializer.deserialize(bytes) as unknown[]);
}

File: sdks/java/src/main/java/org/byteveda/taskito/serialization/Serializer.java
package org.byteveda.taskito.serialization;

import java.lang.reflect.Type;

/** Converts task payloads and results to and from the opaque bytes the core stores. */
public interface Serializer {
    byte[] serialize(Object value);

    <T> T deserialize(byte[] bytes, Class<T> type);

    /**
     * Deserialize to a possibly-generic {@link Type} (from a
     * {@code TypeReference}). The default handles plain {@code Class} types and
     * rejects generic ones; a generics-aware serializer (e.g. {@link JsonSerializer})
     * overrides this.
     */
    default Object deserialize(byte[] bytes, Type type) {
        if (type instanceof Class) {
            return deserialize(bytes, (Class<?>) type);
        }
        throw new UnsupportedOperationException(getClass().getSimpleName()
                + " does not support the generic payload type " + type
                + "; use a Jackson-based serializer or a non-generic Task payload type");
    }

    /**
     * Call-shaped encoding for task payloads. Wire serializers (e.g.
     * {@link CborSerializer}) override this to write the cross-SDK call body
     * {@code [args, kwargs]} from the binding contract; the default keeps the
     * bare-value body. Results always use {@link #serialize}/{@link #deserialize}.
     */
    default byte[] serializeCall(Object payload) {
        return serialize(payload);
    }

    /** Inverse of {@link #serializeCall}: decode a call payload to the handler argument. */
    default Object deserializeCall(byte[] bytes, Type payloadType) {
        return deserialize(bytes, payloadType);
    }
}
